pandas/io/formats/format.py

```diff
@@ -1198,6 +899,209 @@ def _get_column_name_list(self) -> List[str]:
+        line_terminator: Optional[str] = None,
+        chunksize: Optional[int] = None,
+        date_format: Optional[str] = None,
+        doublequote: bool = True,
+        escapechar: Optional[str] = None,
+        errors: str = "strict",
+        storage_options: StorageOptions = None,
+    ) -> Optional[str]:
+        """
+        Render dataframe as comma-separated file.
+        """
+        from pandas.io.formats.csvs import CSVFormatter
+
+        csv_formatter = CSVFormatter(
+            path_or_buf=path_or_buf,
+            line_terminator=line_terminator,
+            sep=sep,
+            encoding=encoding,
+            errors=errors,
+            compression=compression,
+            quoting=quoting,
+            cols=columns,
+            index_label=index_label,
+            mode=mode,
+            chunksize=chunksize,
+            quotechar=quotechar,
+            date_format=date_format,
+            doublequote=doublequote,
+            escapechar=escapechar,
+            storage_options=storage_options,
+            formatter=self.fmt,
+        )
+        csv_formatter.save()
+
+        if path_or_buf is None:
+            assert isinstance(csv_formatter.path_or_buf, StringIO)
+            return csv_formatter.path_or_buf.getvalue()
+
+        return None
+
+    def _get_result(
+        self,
+        string: str,
+        buf: Optional[FilePathOrBuffer[str]] = None,
+        encoding: Optional[str] = None,
+    ) -> Optional[str]:
+        """
+        Perform serialization. Write to buf or return as string if buf is None.
+        """
+        with self._get_buffer(buf, encoding=encoding) as f:
+            f.write(string)
+            if buf is None:
+                return f.getvalue()
+            return None
+
+    @contextmanager
+    def _get_buffer(
+        self, buf: Optional[FilePathOrBuffer[str]], encoding: Optional[str] = None
+    ):
+        """
+        Context manager to open, yield and close buffer for filenames or Path-like
+        objects, otherwise yield buf unchanged.
+        """
+        if buf is not None:
+            buf = stringify_path(buf)
+        else:
+            buf = StringIO()
+
+        if encoding is None:
+            encoding = "utf-8"
+        elif not isinstance(buf, str):
+            raise ValueError("buf is not a file name and encoding is specified.")
+
+        if hasattr(buf, "write"):
+            yield buf
+        elif isinstance(buf, str):
+            with open(buf, "w", encoding=encoding, newline="") as f:
+                # GH#30034 open instead of codecs.open prevents a file leak
+                #  if we have an invalid encoding argument.
+                # newline="" is needed to roundtrip correctly on
+                #  windows test_to_latex_filename
+                yield f
+        else:
+            raise TypeError("buf is not a file name and it has no write method")
+
+
 # ----------------------------------------------------------------------
```

Comment:
would not be averse to splitting this file up if ppossible (e.g. maybe make array.py) for all of the formatters.

Reply:
certainly better as a followon